cmd/daemonset-check-refactor/main.go

```diff
@@ -0,0 +1,286 @@
+
+	// Catch panics.
+	var r interface{}
+	defer func() {
+		r = recover()
+		if r != nil {
+			log.Infoln("Recovered panic:", r)
+			reportErrorsToKuberhealthy([]string{"kuberhealthy/daemonset: " + r.(string)})
+		}
+	}()
+
+	//checkPodTimeout = time.Duration(1) * time.Second
+	// Start daemonset check.
+	runCheck()
+}
+
+
+// setCheckConfigurations sets Daemonset configurations
+func setCheckConfigurations(now time.Time) {
+	hostName = getHostname()
+	daemonSetName = checkDSName + "-" + hostName + "-" + strconv.Itoa(int(now.Unix()))
+}
+
+// listenForInterrupts watches the signal and done channels for termination.
+func listenForInterrupts() {
+
+	// Relay incoming OS interrupt signals to the signalChan
+	signal.Notify(signalChan, os.Interrupt, os.Kill)
+	sig :=<-signalChan
+	log.Infoln("Received an interrupt signal from the signal channel.")
+	log.Debugln("Signal received was:", sig.String())
+
+	go shutdown(doneChan)
+	// wait for checks to be done shutting down before exiting
+	select {
+	case err := <-doneChan:
+		if err != nil {
+			log.Errorln("Error waiting for pod removal during shut down:", err)
+			os.Exit(1)
+		}
+		log.Infoln("Shutdown gracefully completed!")
+		os.Exit(0)
+	case sig = <-signalChan:
+		log.Warningln("Shutdown forced from multiple interrupts!")
+		os.Exit(1)
+	case <-time.After(time.Duration(shutdownGracePeriod)):
+		log.Errorln("Shutdown took too long. Shutting down forcefully!")
+		os.Exit(2)
+	}
+}
+
+// Shutdown signals the DS to begin a cleanup
+func shutdown(sdDoneChan chan error) {
+	shuttingDown = true
+
+	var err error
+	// if the ds is deployed, delete it
+	if daemonSetDeployed {
+
+		log.Infoln("Removing daemonset due to shutdown.")
+
+		err = deleteDS(daemonSetName)
+		if err != nil {
+			log.Errorln("Failed to remove", daemonSetName)
+		}
+
+		daemonSetDeployed = false
+
+		outChan := make(chan error, 10)
+
+		var wg sync.WaitGroup
+		wg.Add(1)
+		go func() {
+			log.Infoln("Worker: waitForPodRemoval started")
+			defer close(outChan)
+			outChan <- waitForPodRemoval()
+			wg.Wait()
+		}()
+
+		select {
+
+		case err = <-outChan:
+			if err != nil {
+				ctxCancel() // cancel the watch context, we have timed out
+				log.Errorln("Error waiting for daemonset pods removal:", err)
+				reportErrorsToKuberhealthy([]string{"kuberhealthy/daemonset: " + err.Error()})
+				sdDoneChan <- err
+				return
+			}
+			log.Infoln("Successfully removed daemonset pods.")
+		case <- time.After(checkPodTimeout):
+			errorMessage := "Reached check pod timeout: " + checkPodTimeout.String() + " waiting for daemonset and daemonset pods removal."
+			log.Errorln(errorMessage)
+			reportErrorsToKuberhealthy([]string{"kuberhealthy/daemonset: " + errorMessage})
+			sdDoneChan <- errors.New(errorMessage)
+			return
+		case <-ctx.Done():
+			// If there is a cancellation interrupt signal.
+			log.Infoln("Canceling removing daemonset and daemonset pods and shutting down from interrupt.")
+			return
+		default:
```

Comment:
wouldn't this always hit the default case? I think there are only three cases here -- your first three w/o default. Especially since you aren't using a for-loop over the select statement, it will just skip straight to the default case